OpenSim/Common/TimeSeriesData.h

```diff
@@ -0,0 +1,183 @@
+* series data in memory for other OpenSim components to consume.
+*/
+
+#include "DataTable.h"
+
+namespace OpenSim {
+//=============================================================================
+//=============================================================================
+/**
+* TimeSeriesData_ is a DataTable where the rows in the table are guaranteed
+* to be sequentially increasing in time. Intended to store trajectories of
+* data (states, controls, points, etc...) as a function of time. Therefore,
+* in addition to rows being accessible by index, they are also accessible
+* by time.
+* The columns of the DataTable are considered to be dependent variables and
+* time is considered the independent variable. Time, therefore, cannot be 
+* accessed as a column of the table. Time is accessed explicitly. 
+* 
+* @author Ajay Seth
+*/
+template<typename T = SimTK::Real>
+class TimeSeriesData_ : public DataTable_<T> {
+public:
+    TimeSeriesData_() = default;
+
+    TimeSeriesData_(const TimeSeriesData_&) = default;
+
+    TimeSeriesData_<T>& operator = (const TimeSeriesData_&) = default;
+
+    /** Copy construct TimeSeriesData from a generic DataTable */
+    TimeSeriesData_(const DataTable_<T>& dt, const std::string& timeLabel = "time") : 
+                DataTable_<T>(dt) {
+        const TimeSeriesData_* tsd =
+            dynamic_cast<const TimeSeriesData_<T>*>(&dt);
+        if (tsd) {
+            // source DataTable is already TimeSeriesData so make this a copy.
+            *this = *tsd;
+        }
+        else if (this->hasColumn(timeLabel)){
+            size_t c = dt.getColumnIndex(timeLabel);
+            if (!isStrictlyIncreasing(dt.getColumn(c)))
+                throw Exception("TimeSeriesData: cannot construct "
+                "from a DataTable without a sequential time column.");
+            _times = dt.getColumn(int(c));
+            this->updColumnLabels().remove(int(c));
+            int nr = int(dt.getNumRows());
+            int nc = int(dt.getNumCols() - 1);
+            this->updAsMatrix().resize(nr, nc) = 
+                dt.getAsMatrix().block(0, 1, nr, nc);
+        }
+        else {
+            throw Exception("TimeSeriesData: cannot construct "
+                "from a DataTable without a time column.");
+        }
+    }
+
+    virtual ~TimeSeriesData_() {}
+
+    /** Time to index options */
+    enum TimeToIndexOption
+    {
+        Before,
+        After,
+        Nearest
+    };
+
+    /** Access the time data as a Vector of Reals */
+    const SimTK::Vector& getTimes() const { return _times; }
+
+    /** Get the time range that the data spans. */
+    SimTK::Vec2 getTimeRange() const { 
+        return SimTK::Vec2(_times[0], _times[_times.size()-1]);
+    }
+
+    /** Update the time data */
+    SimTK::Vector& updTimes() const { return _times; }
+    
+    /** Determine the closest row index for a given time */
+    size_t findRowIndexForTime(const SimTK::Real& time,
```

Comment:
Having jumped through these hoops many times before, I'm really glad that TimeSeriesData will handle this for me.